File: Exercises/Week02-SOLID-Scala/src/week2sensors/FireSensor.scala
package week2sensors

import scala.util.Random


/**
  * Created by nathanhanak on 1/18/17.
  */
class FireSensor(val location: String) extends Sensor {

  private val sensorType: String = "Fire Sensor"
  protected var batteryPercentage : Int = 100

  /**
    *  returns true/false for whether the sensor is triggered or not.
    *  has a 5% chance of returning true
    */
  def isTriggered(): Boolean = {
    checkBatteryLife()
    drainBattery(10)
    if (Random.nextInt(100) < 5) true else false
  }

  protected def checkBatteryLife(): Unit = {
    if (getBatteryPercentage() <= 0) {
      throw new IllegalStateException(s"Battery for ${this.getClass.toString} at ${location} is dead")
    }
  }

  protected def drainBattery(drainAmount:Int): Unit = {
    batteryPercentage -= drainAmount
  }

  /**
    * returns a description of the location of the sensor such as Lobby
    * 1st floor or In the auditorium.
    * @return
    */
  def getLocation(): String = {
    return location
  }

  /**
    * returns a textual description of the sensor type such as "Fire sensor"
    * or "Smoke sensor".
    * @return String describing the type of sensor
    */
  def getSensortype(): String = sensorType


  /**
    * Returns a number between 0-100 where 0 is empty and
    * 100 is fully charged.
    *
    */
  def getBatteryPercentage(): Int = {
    return batteryPercentage
  }

}

File: Exercises/Week02-SOLID-Java/src/week2sensors/FireSensor.java
package week2sensors;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Created by nathanhanak on 1/19/17.
 */
public class FireSensor implements HazardSensor {

    private Double batteryPercentage = 100.0;
    private String sensorType = "Fire Sensor";
    private String location;
    private List<TriggeredResponse> responses;


    public FireSensor(String location, List<TriggeredResponse> responses) {
        this.responses = responses;
        this.location = location;
    }

    @Override
    public boolean isTriggered() {
        if (getBatteryPercentage() < 10) {
            throw new IllegalStateException("Battery too low to perform operation");
        }
        batteryPercentage -= 10;
        double random = ThreadLocalRandom.current().nextDouble(0, 100);
        boolean triggered = ( random < 10);
        if (triggered) {
            executeSensorResponses();
            return true;
        } else {
            return false;
        }
    }

    @Override
    public String getLocation() {
        return location;
    }

    @Override
    public String getSensorType() {
        return sensorType;
    }

    @Override
    public double getBatteryPercentage() {
        return batteryPercentage;
    }

    private void executeSensorResponses() {
        if (!responses.equals(null)) {
            for (TriggeredResponse t : responses) {
                System.out.println(t.performResponse(getSensorType(), getLocation()));
            }
        }
    }
}